### ISSUE TITLE
Memory leak while using ImageDataGenerator flow function

### ISSUE THREAD

**Issue by zhunzhong07** [fchollet/keras#5902]
When I using the ImageDataGenerator flow function to train the model (train_on_batch), the memory will leak. My keras version is 2.0.1, but the 1.2.0 version does not enconter this problem. Code:

    def extract_imagelist(target_path):

         with open(target_path, 'r') as f:
                images, labels = [], []
                for l in f.readlines():
                    l = l.strip('\n').split()
                    images.append(l[0])
                    labels.append(int(l[1]))
         image_len = len(images)
         all_images = np.empty([image_len, 224, 224, 3])
         all_labels = []
         for i in range(image_len):
               img = image.load_img(images[i], target_size=(224, 224))
               img = image.img_to_array(img)
               img = np.expand_dims(img, axis=0)
               img = preprocess_input(img)        
               lab = labels[i]        
               all_images[i, :, :, :] = img
               all_labels.append (lab)
         all_labels = to_categorical(all_labels)
         return all_images, all_labels

    files_list = "train.txt"

    X, Y = extract_imagelist(files_list)

    datagen = ImageDataGenerator(horizontal_flip=True)

    net.compile(optimizer=SGD(lr=0.001, momentum=0.9), loss='categorical_crossentropy')

    for e in range(10):  
          batches = 0
          for X_batch, Y_batch in datagen.flow(X, Y, batch_size=16):
                loss = net.train_on_batch(X_batch, Y_batch)
                if batches % 20 == 0:
    	        print ('Epoch', e, 'iteration:', batches, 'loss:', loss)
                batches += 1
                if batches >= len(X) / 16:
                    break'

**Comment by Namnamseo** [comment]
I'm having some memory leak with `ImageDataGenerator`. I'm using `fit_generator`.

It's strange; I'm running a few different configurations, and some of them leak while others don't. I've switched between multithreading, multiprocessing, and single-thread (the default); it leaks no matter what I choose.

As `keras/image.py` contains nothing but numpy processing, which does not invoke much references, I don't see how this can leak.

**Comment by Namnamseo** [comment]
To be sure, when using a pure `ImageDataGenerator` and the model constructed with pure Keras layers, the memory should not consistently grow after an epoch or two, right? There wouldn't be that much thing to be in memory.

**Comment by zhunzhong07** [comment]
@Namnamseo Did you solved this issue?

**Comment by Namnamseo** [comment]
I haven't.

I tried training each batch manually, measuring memory with `objgraph` before and after the `train_on_batch`.
Install `objgraph` by pip or such, and try:
```
import objgraph

for e in range(10):
    batches = 0
    for X_batch, Y_batch in datagen.flow(X, Y, batch_size=16):
        print('BEFORE')
        objgraph.show_growth()
        loss = net.train_on_batch(X_batch, Y_batch)
        print('AFTER')
        objgraph.show_growth()
        if batches % 20 == 0:
            print ('Epoch', e, 'iteration:', batches, 'loss:', loss)
        batches += 1
        if batches >= len(X) / 16:
            break
```
For me, after the first call, all objects were garbage collected. In other words, nothing came after `AFTER`. So it means that no Python object is leaking. It might be somewhere in the backend(TensorFlow or Theano), or even (but possibly not) CUDA.
